debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotServerReader.java

```diff
@@ -0,0 +1,216 @@
+
+import org.apache.kafka.connect.source.SourceRecord;
+
+import io.debezium.function.BufferedBlockingConsumer;
+import io.debezium.function.Predicates;
+import io.debezium.jdbc.JdbcConnection;
+import io.debezium.relational.TableId;
+import io.debezium.util.Clock;
+import io.debezium.util.Strings;
+
+/**
+ * A component that performs a snapshot of a MySQL server, and records the schema changes in {@link MySqlSchema}.
+ * 
+ * @author Randall Hauch
+ */
+public class SnapshotServerReader extends AbstractSnapshotReader {
+
+    private final boolean includeData;
+    private final SnapshotReaderMetrics metrics;
+    private final SnapshotMySQLUtility snapshotMySQLUtility;
+
+    /**
+     * Create a snapshot reader.
+     * 
+     * @param name the name of this reader; may not be null
+     * @param context the task context in which this reader is running; may not be null
+     */
+    public SnapshotServerReader(String name, MySqlTaskContext context) {
+        super(name, context);
+        this.includeData = !context.isSchemaOnlySnapshot();
+        metrics = new SnapshotReaderMetrics(context.clock());
+        metrics.register(context, logger);
+        snapshotMySQLUtility = new SnapshotMySQLUtility(context);
+    }
+
+    /**
+     * Perform the snapshot using the same logic as the "mysqldump" utility.
+     */
+    protected void execute() {
+        context.configureLoggingContext("snapshot");
+        final AtomicReference<String> sql = new AtomicReference<>();
+        final JdbcConnection mysql = context.jdbc();
+        final MySqlSchema schema = context.dbSchema();
+        final SourceInfo source = context.source();
+        final Clock clock = context.clock();
+        final long ts = clock.currentTimeInMillis();
+        logger.info("Starting snapshot for {} with user '{}'", context.connectionString(), mysql.username());
+        logRolesForCurrentUser(mysql);
+        logServerInformation(mysql);
+        try {
+            metrics.startSnapshot();
+
+            snapshotMySQLUtility.beginTransaction();
+
+            try {
+                // I can't help but feel like there's got to be a better way to do this.
+                if (!isRunning()) return;
+
+                snapshotMySQLUtility.lockAllTables();
+
+                if (!isRunning()) return;
+                //logger.info("\tsnapshot continuing with database(s): {}", includedDatabaseNames);
+
+                snapshotMySQLUtility.maybeReadTableList();
+                List<TableId> tableIds = snapshotMySQLUtility.getTableIds();
+                Set<String> readableDatabaseNames = snapshotMySQLUtility.getReadableDatabaseNames();
+                Map<String, List<TableId>> tableIdsByDbName = snapshotMySQLUtility.getTableIdsByDbName();
+
+
+                // From this point forward, all source records produced by this connector will have an offset that includes a
+                // "snapshot" field (with value of "true").
+
+                // ------
```

Comment:
I like the idea of the flow to be split into method calls but I think it should be done for all steps/operations. As this code iw very complex and it takes time to understand it would be great if we would be able to describe MySQL snapshotting process with high-level names/block and ALL those blocks would be transformed into methods so the execute method will turn into a self-documenting workflow as it will only call those blocks and do just meybe some decision making.

Reply:
Fair enough, I'm also a fan of this complex logic being more readable.  I'll see what other break-downs I can do.